lib/matplotlib/cbook/__init__.py

```diff
@@ -1571,6 +1608,219 @@ def pts_to_prestep(x, *args):
+
+            start = get_interp_point(ind0)
+            end = get_interp_point(ind1)
+        else:
+            # the purpose of the next two lines is for when y2 is a
+            # scalar like 0 and we want the fill to go all the way
+            # down to 0 even if none of the y1 sample points do
+            start = xslice[0], y2slice[0]
+            end = xslice[-1], y2slice[-1]
+
+        X[0] = start
+        X[N + 1] = end
+
+        X[1:N + 1, 0] = xslice
+        X[1:N + 1, 1] = y1slice
+        X[N + 2:, 0] = xslice[::-1]
+        X[N + 2:, 1] = y2slice[::-1]
+
+        if dir == 'y':
+            polys.append(X)
+        elif dir == 'x':
+            polys.append(X.T[::-1].T)
+    return polys
+
+
+def _pts_to_betweenstep(x, *args):
+    """
+    Convert continuous line to between-steps.
+
+    Given a set of ``N`` edges and ``N - 1`` values padded with Nan to size N,
+    converts to ``2N - 2`` points, which when connected linearly give
+    a step function connecting step edges at a specified value
+
+    Parameters
+    ----------
+    x : array
+        The x location of the step edges. May be empty.
+
+    y1, ..., yp : array
+        y arrays to be turned into steps; must have length as ``x`` +/- 1.
+    Returns
+    -------
+    out : array
+        The x and y values converted to steps in the same order as the input;
+        can be unpacked as ``x_out, y1_out, ..., yp_out``.  If the input is
+        length ``N``, each of these arrays will be length ``2N - 2``. For
+        ``N=0``, the length will be 0.
+
+    """
+    args = np.array(args)
+    step_length = max(2 * max(len(x), len(args[0])) - 2, 0)
+    steps = np.zeros((1 + len(args), step_length))
+
+    xlike = len(x) == len(args[0]) + 1
+    ylike = len(x) + 1 == len(args[0])
+
+    def __between__(steps, sl0, sl1, _x, _args):
+        # Be agnostic whether xlike or ylike
+        if _x.flatten().shape != x.shape:
+            if _x.flatten().shape[-1] == _x.shape[-1]:
+                _x = _x[0]
+        steps[sl0, ::steps.shape[-1]-1] = _x[::_x.shape[-1]-1]
+        steps[sl0, 2::2] = _x[1:-1]
+        steps[sl0, 1:-1:2] = _x[1:-1]
+        steps[sl1, 0::2] = _args
+        steps[sl1, 1::2] = _args
+        return steps
+
+    if xlike:
+        steps = __between__(steps, 0, slice(1, None),
+                            x, args)
+    elif ylike:
+        steps = __between__(steps, slice(1, None), 0,
+                                args, x)
+    else:
+        # Fall back to steps-post (e.g. legend drawing)
+        steps = pts_to_poststep(x, *args)
+
+    return steps
+
+
+def _pts_to_betweenstep_edges(x, *args):
+    """
+    Convert continuous line to between-steps, adding edges
+
+    Given a set of ``N`` edges and ``N - 1`` values padded with Nan to size N,
+    converts to ``2N`` points, which when connected linearly give
+    a step function connecting step edges at a specified value, with the first
+    and last edge going to 0
```

Comment:
If you are going to keep `steps_edges`, which we were somewhat ambivalent about on the call, then you probably need to make the "edge" variable, rather than just zero.  As soon as we are talking about a new kwarg, I'm somewhat leaning towards this should have been its own method rather than a new draw style and kwarg:

`ax.draw_hist_steps(bins, vals, bottom=0)` returning a 'histLine2D` that could then be updated in an animation, would make life a lot simpler.  

Reminder that there will be a zoom call today.  Please check the matplotlib gitter if you'd like to call in.

Reply:
There is a strong demand from the HEP community for what drawstyle='edges' does, even if it doesn't end up being done this way.

I see your point that passing a kwarg to step-style which only works with one of them is a bit clumsy. On the other hand, I do think the changes semantically belong to the same group as step and more importantly while the logic can be separated from the current step-styles, it still has to go in somewhere and it's not clear to me what the benefit of this separation should be.

Sorry, I was travelling and not able to connect to the call.